libexec/rtld-elf/aarch64/rtld_start.S

```diff
@@ -101,71 +101,79 @@ END(.rtld_start)
  * x17 = &_rtld_bind_start
  */
 ENTRY(_rtld_bind_start)
-#ifdef __CHERI_PURE_CAPABILITY__
-	/* Currently unused and unimplemented for pure-capability code */
-	brk #0x1
-#else /* defined(__CHERI_PURE_CAPABILITY__) */
 	.cfi_startproc
-	mov	x17, sp
+	mov	PTR(17), PTRN(sp)
 
 	/* Save frame pointer and SP */
-	stp	x29, x30, [sp, #-16]!
-	mov	x29, sp
-	.cfi_def_cfa x29, 16
-	.cfi_offset x30, -8
-	.cfi_offset x29, -16
+	stp	PTR(29), PTR(30), [PTRN(sp), #-(PTR_WIDTH * 2)]!
+	mov	PTR(29), PTRN(sp)
+	.cfi_def_cfa PTR(29), (PTR_WIDTH * 2)
+	.cfi_offset PTR(30), -(PTR_WIDTH * 1)
+	.cfi_offset PTR(29), -(PTR_WIDTH * 2)
 
 	/* Save the arguments */
-	stp	x0, x1, [sp, #-16]!
-	stp	x2, x3, [sp, #-16]!
-	stp	x4, x5, [sp, #-16]!
-	stp	x6, x7, [sp, #-16]!
-	stp	x8, xzr, [sp, #-16]!
+	stp	CAP(0), CAP(1), [PTRN(sp), #-(CAP_WIDTH * 2)]!
+	stp	CAP(2), CAP(3), [PTRN(sp), #-(CAP_WIDTH * 2)]!
+	stp	CAP(4), CAP(5), [PTRN(sp), #-(CAP_WIDTH * 2)]!
+	stp	CAP(6), CAP(7), [PTRN(sp), #-(CAP_WIDTH * 2)]!
+	stp	CAP(8), CAP(zr), [PTRN(sp), #-(CAP_WIDTH * 2)]!
 
 	/* Save any floating-point arguments */
-	stp	q0, q1, [sp, #-32]!
-	stp	q2, q3, [sp, #-32]!
-	stp	q4, q5, [sp, #-32]!
-	stp	q6, q7, [sp, #-32]!
-
-	/* Calculate reloff */
-	ldr	x2, [x17, #0]	/* Get the address of the entry */
-	sub	x1, x2, x16	/* Find its offset */
-	sub	x1, x1, #8	/* Adjust for x16 not being at offset 0 */
-	/* Each rela item has 3 entriesso we need reloff = 3 * index */
+	stp	q0, q1, [PTRN(sp), #-32]!
+	stp	q2, q3, [PTRN(sp), #-32]!
+	stp	q4, q5, [PTRN(sp), #-32]!
+	stp	q6, q7, [PTRN(sp), #-32]!
+
+	/*
+	 * Calculate gotoff. We have &GOT[x + 3] on the stack and &GOT[2],
+	 * which gives x = &GOT[x + 3] - &GOT[2] - 1. We do our calculations
+	 * still scaled by PTR_WIDTH and re-scale to be reloff below.
+	 */
+	ldr	PTR(2), [PTR(17), #0]	/* Get the address of the entry */
+	sub	x1, x2, x16		/* Find its offset from &GOT[2] */
+	sub	x1, x1, #PTR_WIDTH	/* Turn into an offset from &GOT[3] */
+	/*
```

Comment:
```suggestion

	/*
```

Reply:
So, it's a bit anti-style(9), but this is deliberate because the big comment block above still applies to this block of code (so far it's only calculated the address displacement in .got.plt, which needs to be turned into an address displacement in .rela.plt). I guess I could move it all up front.

Reply:
I ended up just merging this with the comment above.